pkg/teeth/bigquery.go

```diff
@@ -0,0 +1,167 @@
+	PullRequestURL string             `bigquery:"pull_request_html_url"`
+	ProcessedAt    time.Time          `bigquery:"processed_at"`
+	CommentID      bigquery.NullInt64 `bigquery:"comment_id"`
+	Status         string             `bigquery:"status"`
+	JobName        string             `bigquery:"job_name"`
+}
+
+// BigQuery provides a client to BigQuery API.
+type BigQuery struct {
+	config *BQConfig
+	client *bigquery.Client
+}
+
+// NewBigQuery creates a new instance of a BigQuery client with config.
+func NewBigQuery(ctx context.Context, config *BQConfig) (*BigQuery, error) {
+	c, err := bigquery.NewClient(ctx, config.ProjectID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create bigquery.Client: %w", err)
+	}
+	return &BigQuery{
+		config: config,
+		client: c,
+	}, nil
+}
+
+// Close closes the BigQuery client.
+func (bq *BigQuery) Close() error {
+	if err := bq.client.Close(); err != nil {
+		return fmt.Errorf("failed to close client: %w", err)
+	}
+	return nil
+}
+
+// Config returns the BigQuery client config.
+func (bq *BigQuery) Config() *BQConfig {
+	return bq.config
+}
+
+// QueryLatest executes the source query for the latest PublisherSourceRecords
+// to process.
+func (bq *BigQuery) QueryLatest(ctx context.Context) ([]*PublisherSourceRecord, error) {
+	qStr, err := populatePublisherSourceQuery(ctx, bq.config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to populate query: %w", err)
+	}
+
+	// below copied from https://pkg.go.dev/cloud.google.com/go/bigquery#hdr-Querying
+	q := bq.client.Query(qStr)
+	it, err := q.Read(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read: %w", err)
+	}
+	results := make([]*PublisherSourceRecord, 0)
+	for {
+		r := &PublisherSourceRecord{}
+		err := it.Next(r)
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to read result: %w", err)
+		}
+		results = append(results, r)
+	}
+	return results, nil
+}
+
+func populatePublisherSourceQuery(ctx context.Context, config *BQConfig) (string, error) {
+	type Tables struct {
+		PullRequestEventsTable       string
+		InvocationCommentStatusTable string
+		EventsTable                  string
+		LeechStatusTable             string
+	}
+	tablePrefix := fmt.Sprintf("%s.%s.", config.ProjectID, config.DatasetID)
+	fullTableNames := &Tables{
+		PullRequestEventsTable:       tablePrefix + config.PullRequestEventsTable,
+		InvocationCommentStatusTable: tablePrefix + config.InvocationCommentStatusTable,
+		EventsTable:                  tablePrefix + config.EventsTable,
+		LeechStatusTable:             tablePrefix + config.LeechStatusTable,
+	}
+	tmpl, err := template.New("publisher").Parse(PublisherSourceQuery)
+	if err != nil {
+		return "", fmt.Errorf("failed to set up sql template: %w", err)
+	}
+	var b bytes.Buffer
+	if err = tmpl.Execute(&b, fullTableNames); err != nil {
+		return "", fmt.Errorf("failed to execute sql template: %w", err)
+	}
```

Comment:
You can cheat a bit here if you want:

```suggestion
	tablePrefix := fmt.Sprintf("%s.%s.", config.ProjectID, config.DatasetID)
	tmpl, err := template.New("publisher").Parse(PublisherSourceQuery)
	if err != nil {
		return "", fmt.Errorf("failed to set up sql template: %w", err)
	}
	var b bytes.Buffer
	if err = tmpl.Execute(&b, map[string]string{
		"PullRequestEventsTable":       tablePrefix + config.PullRequestEventsTable,
		"InvocationCommentStatusTable": tablePrefix + config.InvocationCommentStatusTable,
		"EventsTable":                  tablePrefix + config.EventsTable,
		"LeechStatusTable":             tablePrefix + config.LeechStatusTable,
	}); err != nil {
		return "", fmt.Errorf("failed to execute sql template: %w", err)
	}
```

Reply:
Nice! Thanks!